### DISCUSSION TITLE
About Stats

### DISCUSSION THREAD

**Discussion by mezut35** [blockscout/blockscout#9280] [Support]
Hello, I am attempting a trial with Blockscout, but no matter what I do, I can't get the stats service to run. Other transactions are coming through, and blocks are being received, but when I click on the stats and chart menu, no data comes through. I used your example Docker Compose file and entered the details of my EVM server. I can't understand what I need to do for the stats section. I would appreciate your assistance with this.

![image](https://github.com/blockscout/blockscout/assets/114423949/acf09aff-544e-420c-9c91-9f204b074111)

version: '3.9'

services:
  redis_db:
    extends:
      file: ./services/redis.yml
      service: redis_db

  db-init:
    extends:
      file: ./services/db.yml
      service: db-init

  db:
    extends:
      file: ./services/db.yml
      service: db

  backend:
    depends_on:
      - db
      - redis_db
    extends:
      file: ./services/backend.yml
      service: backend
    build:
      context: ..
      dockerfile: ./docker/Dockerfile
      args:
        CACHE_EXCHANGE_RATES_PERIOD: "false"
        API_V1_READ_METHODS_DISABLED: "false"
        DISABLE_WEBAPP: "false"
        API_V1_WRITE_METHODS_DISABLED: "false"
        CACHE_TOTAL_GAS_USAGE_COUNTER_ENABLED: ""
        CACHE_ADDRESS_WITH_BALANCES_UPDATE_INTERVAL: ""
        ADMIN_PANEL_ENABLED: "false"
        RELEASE_VERSION: 6.0.0
    links:
      - db:database
    environment:
        ETHEREUM_JSONRPC_HTTP_URL: http://192.168.0.124:8545/
        ETHEREUM_JSONRPC_TRACE_URL: http://192.168.0.124:8545/
        ETHEREUM_JSONRPC_WS_URL: ws://192.168.0.124:8546/
        CHAIN_ID: '500'

  visualizer:
    extends:
      file: ./services/visualizer.yml
      service: visualizer

  sig-provider:
    extends:
      file: ./services/sig-provider.yml
      service: sig-provider

  frontend:
    depends_on:
      - backend
    extends:
      file: ./services/frontend.yml
      service: frontend
    environment:
      # NEXT_PUBLIC_OTHER_LINKS: ""
      NEXT_PUBLIC_FOOTER_LINKS: "https://raw.githubusercontent.com/blockscout/frontend-configs/main/configs/footer-links/base-goerli.json"
      # NEXT_PUBLIC_NAVIGATION_HIDDEN_LINKS: ""
      # NEXT_PUBLIC_FEATURED_NETWORKS: ""
      NEXT_PUBLIC_OG_DESCRIPTION: "xxxx
      NEXT_PUBLIC_OG_IMAGE_URL: "
      NEXT_PUBLIC_APP_PROTOCOL: "https"
      NEXT_PUBLIC_APP_HOST: "explorertest.test.com"
      NEXT_PUBLIC_API_PROTOCOL: "https"
      NEXT_PUBLIC_API_WEBSOCKET_PROTOCOL: "wss"
      NEXT_PUBLIC_API_HOST: "explorertest.test.com"
      NEXT_PUBLIC_NETWORK_NAME: "test"
      NEXT_PUBLIC_NETWORK_SHORT_NAME: "test"
      NEXT_PUBLIC_NETWORK_ID: 571
      NEXT_PUBLIC_NETWORK_RPC_URL: https://rpc.xxxx.com/
      NEXT_PUBLIC_NETWORK_CURRENCY_NAME: "test"
      NEXT_PUBLIC_NETWORK_CURRENCY_SYMBOL: "test"
      NEXT_PUBLIC_NETWORK_VERIFICATION_TYPE: "validation"
      NEXT_PUBLIC_IS_TESTNET: "false"
      NEXT_PUBLIC_HOMEPAGE_CHARTS: "['daily_txs']"
      NEXT_PUBLIC_NETWORK_LOGO: 
      NEXT_PUBLIC_NETWORK_LOGO_DARK: 
      NEXT_PUBLIC_NETWORK_ICON: 
      NEXT_PUBLIC_NETWORK_ICON_DARK: 
      NEXT_PUBLIC_VIEWS_ADDRESS_IDENTICON_TYPE: "blockie"
      # NEXT_PUBLIC_VIEWS_NFT_MARKETPLACES: ""
      # NEXT_PUBLIC_IS_ACCOUNT_SUPPORTED: "true"
      NEXT_PUBLIC_WALLET_CONNECT_PROJECT_ID: 
      NEXT_PUBLIC_AD_BANNER_PROVIDER: "none"
      NEXT_PUBLIC_AD_TEXT_PROVIDER: "none"

  stats-db-init:
    extends:
      file: ./services/stats.yml
      service: stats-db-init

  stats-db:
    depends_on:
      - backend
    extends:
      file: ./services/stats.yml
      service: stats-db

  stats:
    depends_on:
      - stats-db
    extends:
      file: ./services/stats.yml
      service: stats

  proxy:
    depends_on:
      - backend
      - frontend
      - stats
    extends:
      file: ./services/nginx.yml
      service: proxy

**Comment by GimluCom** [comment]
Have you read through https://docs.blockscout.com/for-developers/configuration-options/charts-and-stats,
and is your Blockscout instance fully indexed?

**Reply by mezut35** [reply]
I am working on my own private EVM network that I set up. How can I determine if indexing has been fully completed? Additionally, there are error logs in the Blockscout container like the ones below.. Could this be related?
Blockscout has been running for about 8-10 hours. Transactions, blocks, and top account data appear and are being updated on the webpage. localhost:8080/api*** actually returns some results, but the charts and stats section of the webpage says "Something went wrong. Try refreshing the page or come back later."

I suspect the following logs may be related to this issue:

User
{"time":"2024-01-29T19:53:04.958Z","severity":"error","message":"failed to fetch internal transactions for blocks: ** (ErlangError) Erlang error: [%{code: -32000, data: %{block_number: 233189, transaction_hash: \"0x959e9a4726a8e44e14245964ebc9ccdc13d5c88c427b56cd3e8097965be64873\", transaction_index: 0}, message: \"required historical state unavailable (reexec=128)\"}]","metadata":{"count":10,"error_count":10,"fetcher":"internal_transaction"}}
{"time":"2024-01-29T19:53:04.965Z","severity":"error","message":"failed to fetch internal transactions for blocks: ** (ErlangError) Erlang error: [%{code: -32000, data: %{block_number: 233187, transaction_hash: \"0xbb7db5a4ef2536a84de03078927efdba2b3a2ffd89f642e9784302a47b223057\", transaction_index: 0}, message: \"required historical state unavailable (reexec=128)\"}]","metadata":{"count":10,"error_count":10,"fetcher":"internal_transaction"}}
{"time":"2024-01-29T19:53:04.967Z","severity":"error","message":"failed to fetch internal transactions for blocks: ** (ErlangError) Erlang error: [%{code: -32000, data: %{block_number: 233150, transaction_hash: \"0xae7e33ea91dcad46df774a695d03c7f9160daa0cf03dc238add0d69aeef63f2b\", transaction_index: 0}, message: \"required historical state unavailable (reexec=128)\"}]","metadata":{"count":10,"error_count":10,"fetcher":"internal_transaction"}}
{"time":"2024-01-29T19:53:04.969Z","severity":"error","message":"failed to fetch internal transactions for blocks: ** (ErlangError) Erlang error: [%{code: -32000, data: %{block_number: 345862, transaction_hash: \"0xbb383d0de99940146ee70883bb09501cb3c0e889820cee672f02996777406fb7\", transaction_index: 0}, message: \"required historical state unavailable (reexec=128)\"}]","metadata":

**Reply by GimluCom** [reply]
You would need to make sure your RPC node is an archive node. If you don't want internal transactions you could instead disable them with ENV `INDEXER_DISABLE_INTERNAL_TRANSACTIONS_FETCHER=true`

**Comment by mezut35** [comment]
also stats logs

2024-01-30T07:00:00.045376Z  INFO stats::charts::updater: running full update due to lack of history data min_blockscout_block=0 chart="newVerifiedContracts"
2024-01-30T07:00:00.047422Z  INFO stats_server::update_service: successfully updated chart chart="verifiedContractsGrowth"
2024-01-30T07:00:00.047446Z  INFO stats_server::update_service: scheduled next run of chart update in 86399.952562635s chart="verifiedContractsGrowth"
2024-01-30T07:00:00.388601Z  INFO stats_server::update_service: successfully updated chart chart="txnsFee"
2024-01-30T07:00:00.388645Z  INFO stats_server::update_service: scheduled next run of chart update in 86399.611366075s chart="txnsFee"
2024-01-30T08:00:00.001179Z  INFO stats_server::update_service: updating chart chart="newBlocks"
2024-01-30T08:00:00.001259Z  INFO stats_server::update_service: updating chart chart="contractsGrowth"
2024-01-30T08:00:00.018805Z  INFO stats::charts::updater::dependent: updating parent chart_name="contractsGrowth" parent_chart_name="newContracts"
2024-01-30T08:00:00.021668Z  INFO stats::charts::updater: running full update due to min blocks mismatch min_blockscout_block=0 min_chart_block=438365 chart="newBlocks"
2024-01-30T08:00:00.021943Z  INFO stats::charts::updater: running full update due to min blocks mismatch min_blockscout_block=0 min_chart_block=438365 chart="newContracts"
2024-01-30T08:00:00.021956Z  INFO stats::charts::updater::batch: start batch update last_row=None
2024-01-30T08:00:00.023220Z  INFO stats::charts::updater::batch: run 1/2 step of batch update from=2023-12-18 to=2024-01-17
2024-01-30T08:00:00.029407Z  INFO stats::charts::updater::batch: 1/2 step of batch done found=6 elapsed=6.16155ms
2024-01-30T08:00:00.031266Z  INFO stats::charts::updater::batch: run 2/2 step of batch update from=2024-01-17 to=2024-02-16
2024-01-30T08:00:00.032930Z  INFO stats::charts::updater::batch: 2/2 step of batch done found=6 elapsed=1.636853ms
2024-01-30T08:00:00.036591Z  INFO stats_server::update_service: successfully updated chart chart="contractsGrowth"
2024-01-30T08:00:00.036615Z  INFO stats_server::update_service: scheduled next run of chart update in 86399.963393963s chart="contractsGrowth"
2024-01-30T08:00:00.197391Z  INFO stats_server::update_service: successfully updated chart chart="newBlocks"
2024-01-30T08:00:00.197416Z  INFO stats_server::update_service: scheduled next run of chart update in 86399.80259207s chart="newBlocks"

**Comment by mezut35** [comment]
![image](https://github.com/blockscout/blockscout/assets/114423949/0c79a431-3299-4e87-8a42-dbb1657ddb5e)
![image](https://github.com/blockscout/blockscout/assets/114423949/08635210-c989-4a2d-91a5-51e0f0d49823)
![image](https://github.com/blockscout/blockscout/assets/114423949/76a62156-fb18-4d8a-a746-861deaa2641c)

NEXT_PUBLIC_STATS_API_HOST: http://localhost:8080  

![image](https://github.com/blockscout/blockscout/assets/114423949/5f1a730a-5eb7-45c2-9a96-f3861c797dea)



When I try localhost:8080, I do not receive any response from the browser. However, when I type localhost:8080/api/v1/counters, I can access some data. Despite there being no errors in the services and logs, the charts and stats data are not coming through. I couldn't succeed. Could you please help me?

**Reply by PRATHAM-SPS** [reply]
Console :- Access to fetch at 'http://localhost:8080/api/v1/lines' from origin 'http://172.18.0.10:3000' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header has a value 'http://localhost' that is not equal to the supplied origin. Have the server send the header with a valid value, or, if an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

I am facing this error as I click on Charts and Stats . Does anyone have solution for it?

**Reply by GimluCom** [reply]
[Configure CORS](https://github.com/blockscout/blockscout/blob/v6.2.2-beta/docker-compose/proxy/default.conf.template#L41C47-L41C63) like this `add_header 'Access-Control-Allow-Origin' 'http://172.18.0.10:3000' always;`

**Reply by PRATHAM-SPS** [reply]
Thank you @GimluCom :100:

**Reply by PRATHAM-SPS** [reply]
Hey @GimluCom  
console:- stats:1 Access to fetch at 'http://localhost/api/v2/stats' from origin 'http://172.18.0.10:3000' has been blocked by CORS policy: Request header field updated-gas-oracle is not allowed by Access-Control-Allow-Headers in preflight response.

I have tried adding updated-gas-oracle in the allowed headers in the above mentioned file by you

How can I solve this error?

**Reply by GimluCom** [reply]
Have you set `NEXT_PUBLIC_APP_HOST`? https://github.com/blockscout/frontend/blob/main/docs/ENVS.md

**Reply by PRATHAM-SPS** [reply]
yes its set to localhost it was an default value

**Reply by GimluCom** [reply]
Try setting the correct IP/domain in `NEXT_PUBLIC_APP_HOST` and `NEXT_PUBLIC_API_HOST`.
Also you should access your blockscout instance through the nginx proxy instead of accessing the frontend directly.

**Reply by PRATHAM-SPS** [reply]
I have the whole setup inside the docker frontend,backend as well as microservices, while running on local host  ip changes from http://172.18.0.10:3000/ to http://172.18.0:9:3000/ again and again as i recreate / start or stop the container.

**Reply by GimluCom** [reply]
The nginx proxy is available at port 80 (http://127.0.0.1 or http://localhost)

**Reply by PRATHAM-SPS** [reply]
At http://localhost/ everything is working perfectly, I want to make it work on this ip address 172.18.0.10 or any other ip, any idea how can i do it?

only one error I am facing **console:- stats:1 Access to fetch at 'http://localhost/api/v2/stats' from origin 'http://172.18.0.10:3000' has been blocked by CORS policy: Request header field updated-gas-oracle is not allowed by Access-Control-Allow-Headers in preflight response.** , other things are working perfectly here .

**Reply by GimluCom** [reply]
```
NEXT_PUBLIC_APP_HOST=127.0.0.1
NEXT_PUBLIC_API_HOST=127.0.0.1
```
Then go to http://127.0.0.1

**Reply by PRATHAM-SPS** [reply]
yes its working fine on this IP address as its an localhost , can we make it work on http://172.18.0.10:3000/ this address??

**Reply by GimluCom** [reply]
Could you please explain why? What are you trying to accomplish?

**Reply by PRATHAM-SPS** [reply]
![image](https://github.com/blockscout/blockscout/assets/94799826/13fcb4da-7d30-46c2-9b9b-61c6ae7e0d3d)
**info at http://localhost/**




...................................................xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx......................................................................
![image](https://github.com/blockscout/blockscout/assets/94799826/1bb60539-9394-4c22-ac15-980150bc6645)
**info at  http://172.18.0.10:3000/**

I am  not able to access this stats info, if the frontend is on any other ip address rather than localhost or 127.0.0.1

**Reply by GimluCom** [reply]
Find your internal IP with `hostname -I` and use that.

**Reply by PRATHAM-SPS** [reply]
Ok , Thankyou @GimluCom ✨ , I will try it and revert back to you again.

**Reply by khiem20tc** [reply]
The same issue is with the local and host IP, could u tell me fixed successfully?
